### PR TITLE
NIM-3037 - Add deselect all pages logic to selection mixin, add select all pages event

### PR DESCRIPTION
**Pull request by TristanDevos**
https://desire2learn.atlassian.net/browse/NIM-3037

These changes are related to [this thread](https://d2l.slack.com/archives/C036FJMD4P7/p1720029143526439). These changes do not resolve all issues with the selection-mixin and lazy loading, but gets us to an "okay" state until a more permanent solution is developed. 

1. As a consumer we need to know if the `'none'` state came from deselect-all (the [selection-select-all](https://github.com/BrightspaceUI/core/blob/main/components/selection/selection-select-all.js) component) or from deselecting the last loaded list item.
   - ex. we have 4 items that are selected, 2 are loaded, 2 are not. We deselect the two loaded items so the state becomes 'none', but we don't want to deselect the two unloaded items.
   - In the above scenario, instead of deselecting the two loaded items, we use the `selection-select-all` component to deselect-all. In this case we do want to deselect the two unloaded items. 
   - Now we have a way to know the difference between the two scenarios.

2. As a consumer we need to know when "Select All x Items" ([selection-select-all-pages](https://github.com/BrightspaceUI/core/blob/main/components/selection/selection-select-all-pages.js)) is pressed, even if no selection changes happen
   - This is so that we can toggle any unloaded items, even if all the loaded items are already toggled.
   - The best way I could think of doing this was through an event, but let me know if there is a better idea.

TODO: see if any tests can be updated/added

### CODE REVIEW

File: components/selection/selection-mixin.js [hunk 1]
```diff
@@ -12,11 +12,16 @@ const keyCodes = {
...
-	constructor(keys, state) {
+	constructor(keys, state, allPages) {
 		if (!keys) keys = [];
 		if (!state) state = SelectionInfo.states.none;
 		this._keys = keys;
 		this._state = state;
+		this._allPages = allPages;
+	}
+
+	get allPages() {
```
Comment (TristanDevos) [inline components/selection/selection-mixin.js:23]: I initially had another state for deselect all pages (`noPages`) but went with a new property to signify if the `none` or `all` state applies to all pages. It wouldn't make sense for this to be true when in the `some` state, but I don't think that can happen.

I believe the activity-evaluation-editor (the component I'm modifying this for) is the only consumer of the `select-all-pages-allowed` attribute so it's okay that I remove the `allPages` state. But let me know if you find a reference to it somewhere else.

**Comment by dbatiste** [comment]
Hey Tristan, thanks for raising this use case. It has come up once or twice over the past couple of years, albeit with perhaps slightly different use cases and design. I think we need to have some discussion with design to figure out how we want to support persistent selection before we make any significant or breaking changes to how selection is tracked or exposed on the API. I don't want to discourage contribution, but I also want to be up front about the changes we'll be able to take in the short-term. 😄

**Comment by TristanDevos** [comment]
> Hey Tristan, thanks for raising this use case. It has come up once or twice over the past couple of years, albeit with perhaps slightly different use cases and design. I think we need to have some discussion with design to figure out how we want to support persistent selection before we make any significant or breaking changes to how selection is tracked or exposed on the API. I don't want to discourage contribution, but I also want to be up front about the changes we'll be able to take in the short-term. 😄

Are you saying you'd like to go through design and put in a more permanent solution instead of my interim solution?

**Comment by dbatiste** [comment]
> Are you saying you'd like to go through design and put in a more permanent solution instead of my interim solution?

Yeah, from a core perspective, we need to go through those use-cases, review the design, before we can make any significant changes here.

**Comment by TristanDevos** [comment]
> Yeah, from a core perspective, we need to go through those use-cases, review the design, before we can make any significant changes here.

I was hoping to get this to a semi-working state for our August GA, but I understand.
Would it be possible to just add the event being dispatched from the select-all-pages button? That would still vastly improve the user's experience. Let me know what you think.

**Comment by TristanDevos** [comment]
Discussed in a call with Dave B. and Jeff G. that we'll be waiting for a longer term solution. For now we're just going to increase the default paging on our list to hopefully catch most cases.

**Pull request closed without merging**